File: src/utils/CircleList.java
package utils;

public class CircleList<T> {

    /*Node class*/
    private class NodeList {
        protected NodeList previous;
        private T data;
        protected NodeList next;

        public NodeList() {
            this.previous = null;
            this.next = null;
        }

        public NodeList(T received) {
            this.previous = null;
            this.data = received;
            this.next = null;
        }

        protected void setInformation(T received) {
            this.data = received;
        }

        public Object getData() {
            return this.data;
        }
    }

    private NodeList current;
    private NodeList first;
    private NodeList last;

    /*Constructor*/
    public CircleList() {
        /*this.current = new NodeList();*/
        this.first = this.current;
        this.last = this.current;
        this.current.next = this.current;
        this.current.previous = this.current;
    }

    public CircleList(T received) {
        this.current = new NodeList(received);
        this.first = this.current;
        this.last = this.current;
        this.current.next = this.current;
        this.current.previous = this.current;
    }

    public boolean setInformation(T received) {
        if (current == null)
            return false;
        this.current.setInformation(received);
        return true;
    }

    public Object getData() {
        if (current == null)
            return null;
        return this.current.getData();
    }

    public boolean next() {
        if (current == null)
            return false;
        this.current = this.current.next;
        return true;
    }

    public boolean previous() {
        if (current == null)
            return false;
        this.current = this.current.previous;
        return true;
    }

    public Object inNext() {
        if (current == null)
            return null;
        return this.current.next.getData();
    }

    public Object inPrevious() {
        if (current == null)
            return null;
        return this.current.previous.getData();
    }

    private void makingNode(NodeList receivedNode) {
        if (this.current == null) {
            this.current = receivedNode;
            this.first = this.current;
            this.last = this.current;
            this.current.next = this.current;
            this.current.previous = this.current;
        } else {
            receivedNode.next = this.current;
            receivedNode.previous = this.current.previous;
            this.current.previous.next = receivedNode;
            this.current.previous = receivedNode;
            this.last = receivedNode;
            this.current = receivedNode;
        }
    }

    public void add() {
        makingNode(new NodeList());
    }

    public void add(T received) {
        makingNode(new NodeList(received));
    }

    public boolean removeNode() {
        if (this.current == null)
            return false;

        this.current.previous.next = this.current.next;
        this.current.next.previous = this.current.previous;
        if (this.current == this.first) {
            this.first = this.current.next;
            this.current = this.current.next;
        }
        if (this.current == this.last) {
            this.last = this.current.previous;
            this.current = this.current.previous;
        }
        if (this.current.next == this.current) {
            this.current = null;
            this.first = null;
            this.last = null;
        }
        return true;
    }

    public void deleteList() {
        while (this.removeNode()) ;
    }
}

File: src/utils/CircleList.kt
package utils

class CircleList<T> {
    /*Node class*/
    private inner class NodeList<K>(private var data: K) {
        var previous: NodeList<K>? = null
        var next: NodeList<K>? = null

        fun setInformation(received: K) {
            this.data = received
        }

        fun getData(): K {
            return this.data
        }
    }

    private var current: NodeList<T>? = null
    private var first: NodeList<T>?
    private var last: NodeList<T>?

    /*Constructor*/
    constructor() {
        /*this.current = new NodeList();*/
        first = current
        last = current
        current?.next = current
        current?.previous = current
    }

    constructor(received: T) {
        current = NodeList(received)
        first = current
        last = current
        current?.next = current
        current?.previous = current
    }

    fun setInformation(received: T): Boolean {
        if (current == null) return false
        current?.setInformation(received)
        return true
    }

    val data: T?
        get() = current?.getData()

    operator fun next(): Boolean {
        if (current == null) return false
        current = current?.next
        return true
    }

    fun previous(): Boolean {
        if (current == null) return false
        current = current?.previous
        return true
    }

    fun inNext(): T? {
        return current?.next?.getData()
    }

    fun inPrevious(): T? {
        return current?.previous?.getData()
    }

    private fun makingNode(receivedNode: NodeList<T>) {
        if (current == null) {
            current = receivedNode
            first = current
            last = current
            current?.next = current
            current?.previous = current
        } else {
            receivedNode.next = current
            receivedNode.previous = current?.previous
            current?.previous?.next = receivedNode
            current?.previous = receivedNode
            last = receivedNode
            current = receivedNode
        }
    }

    fun add(received: T) {
        makingNode(NodeList(received))
    }

    fun remove(): Boolean {
        if (current == null) return false
        current?.previous?.next = current?.next
        current?.next?.previous = current?.previous
        if (current === first) {
            first = current?.next
            current = current?.next
        }
        if (current === last) {
            last = current?.previous
            current = current?.previous
        }
        if (current?.next === current) {
            current = null
            first = null
            last = null
        }
        return true
    }

    fun deleteList() {
        while (remove());
    }
}
